oak_client/src/interceptors/mod.rs

```diff
@@ -0,0 +1,57 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use tonic::{Request, Status};
+
+#[cfg(test)]
+mod tests;
+
+pub mod auth;
+pub mod label;
+
+/// Trait for interceptors, to facilitate combining until
+/// https://github.com/hyperium/tonic/issues/499 offers a better solution.
+pub trait Interceptor: Send + Sync + 'static {
+    fn process(&self, request: Request<()>) -> Result<Request<()>, Status>;
+}
+
+pub struct CombinedInterceptor<A: Interceptor, B: Interceptor> {
+    interceptor_a: A,
+    interceptor_b: B,
+}
+
+pub fn combine<A: Interceptor, B: Interceptor>(
+    interceptor_a: A,
+    interceptor_b: B,
+) -> CombinedInterceptor<A, B> {
+    CombinedInterceptor {
+        interceptor_a,
+        interceptor_b,
+    }
+}
+
+impl<A: Interceptor, B: Interceptor> Interceptor for CombinedInterceptor<A, B> {
+    fn process(&self, request: Request<()>) -> Result<Request<()>, Status> {
+        let request = self.interceptor_a.process(request)?;
+        self.interceptor_b.process(request)
+    }
+}
+
+impl<A: Interceptor, B: Interceptor> Into<tonic::Interceptor> for CombinedInterceptor<A, B> {
```

Comment:
Is not it possible to implement `From` here? It is advised to implement `From` instead of `Into`.

Reply:
No, `tonic::Interceptor` is a foreign type, so I cannot implement traits on it.